aiida_aurora/utils/cycling_analysis.py

```diff
@@ -1,144 +1,356 @@
-        for g in below_groups:
-            if g > consecutive_cycles:
-                print(f'Below threshold for {g} cycles!')
-    return data
-
-
-def cycling_analysis(calcjob_node, retrieve_monitor_params=False, consecutive_cycles=2, threshold=0.8, discharge=True):
-    """Perform the cycling analysis. You can provide either the cycler or the monitor calcjob.
-    First, it will try to find and analyse any output of the cycler calcjob.
-    If this does not succeed, the monitor calcjob outputs will be analysed (the results or the last snapshot).
-
-      retrieve_monitor_params :  if True, try to load the monitor parameters from the inputs
+    Used by the frontend Aurora app for plotting.
+
+    The analysis report attached to each plot series includes:
+    - details of the associated monitor, if one was assigned
+    - summary/statistics of the post-processed data
+
+    Parameters
+    ----------
+    `node` : `CalcJobNode`
+        The calculation `node`.
+
+    Returns
+    -------
+    `Tuple[dict, str]`
+        Post-processed data and an analysis report.
+
+    Raises
+    ------
+    `TypeError`
+        If `node` is not a `BatteryCyclerExperiment`.
     """
 
-    monitor_calcjob = None
-    if calcjob_node.process_type == 'aiida.calculations:aurora.cycler':
-        calcjob = calcjob_node
-        if calcjob.get_extra('monitored', False):
-            # find last monitor, if existing
-            qb = QueryBuilder()
-            qb.append(RemoteData, filters={'uuid': calcjob.outputs.remote_folder.uuid}, tag='rf')
-            qb.append(
-                CalcJobNode,
-                with_incoming='rf',
-                edge_filters={'label': 'monitor_folder'},
-                project=['*', 'id'],
-                tag='mon'
-            )
-            qb.order_by({'mon': {'id': 'desc'}})
-            monitor_calcjob = qb.first()[0] if qb.count() else None
-    elif calcjob_node.process_type == 'aiida.calculations:calcmonitor.calcjob_monitor':
-        monitor_calcjob = calcjob_node
-        calcjob = monitor_calcjob.inputs.monitor_folder.get_incoming().get_node_by_label('remote_folder')
-    else:
-        raise TypeError('calcjob_node should be a BatteryCyclerExperiment or a CalcjobMonitor')
+    if node.process_type != "aiida.calculations:aurora.cycler":
+        raise TypeError("`node` is not a `BatteryCyclerExperiment`")
 
-    if monitor_calcjob:
-        print(f"Monitored CalcJob:   <{calcjob.id}> '{calcjob.label}'")
-        print(f"Monitor CalcJob:     <{monitor_calcjob.id}> '{monitor_calcjob.label}'")
+    report = f"CalcJob:   <{node.pk}> '{node.label}'\n"
+
+    sample: BatterySample = node.inputs.battery_sample
+    report += f"Sample:    {sample.label}\n"
+
+    if node.get_extra("monitored", False):
+        report += "Monitored: True\n"
+        monitors = get_monitors(node)
+        report += add_monitor_details(monitors)
```

Comment:
The `monitored` extra gets set by the aiidalab interface right? Perhaps as long as we keep this here (referencing the conversation about possibly moving parts of this to the aiidalab package) we may not want to rely on behavior of a package we are not depending on.

The method `get_monitors` already checks for monitors and returns `{}` if it doesn't find any. So instead of having the inconsistency of `monitored` + empty dict reported as a warning in `add_monitor_details`, I think we could just use this empty dict as the check to decide if it monitored or not. The consistency between that and the presence of the "monitored" `extra` of the node should probably be checked somewhere in the aiidalab package.

Reply:
Well, once the cycling workflow is merged, `monitored` will be set by the workflow after queuing each process.

```
    def run_cycling_step(self):
        """Run the next cycling step."""
        current_keyname = self.worksteps_keynames.pop(0)

        inputs = {
            'code': self.inputs.tomato_code,
            'battery_sample': self.inputs.battery_sample,
            'protocol': self.inputs.protocols[current_keyname],
            'control_settings': self.inputs.control_settings[current_keyname],
        }

        has_monitors = current_keyname in self.inputs.monitor_settings

        if has_monitors:
            inputs['monitors'] = self.inputs.monitor_settings[current_keyname]

        running = self.submit(CyclerCalcjob, **inputs)

        if has_monitors:
            running.set_extra('monitored', True)
        else:
            running.set_extra('monitored', False)

        self.report(f'launching CyclerCalcjob<{running.pk}>')
        return ToContext(subprocesses=append_(running))
```

Reply:
Okay, I see the inconsistency of `monitored` but empty. You're right, there is no need to rely on `monitored`. How about...

```
    report += "Monitored: "

    if monitors := get_monitors(node):
        report += "True\n"
        report += add_monitor_details(monitors)
    else:
        report += "False\n"
```

Reply:
Okay that works. Added a missing check on if `"remote_folder"` is in `node.outputs` 👍